### PR TITLE
SUPNXP-14677 CSV import doc type should be optional for updates

### PR DESCRIPTION
**Pull request by ron1**

### CODE REVIEW

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterDocumentFactory.java [hunk 1]
```diff
@@ -36,6 +36,16 @@ public void createDocument(CoreSession session, String parentPath, String name,
...
+    /**
+     * @return {@code true} if a document with the specified parentPath and name exists. {@code false} otherwise.
+     * @since 8.1
+     */
+    public boolean exists(CoreSession session, String parentPath, String name);
+
+    /**
+     * @deprecated since 8.1
+     */
+    @Deprecated
```
Comment (troger) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterDocumentFactory.java]: Why deprecating this method and add a new one without `values`?

We may not use `values` in the default implementation but the point here is to allow users creating their own factory to make use of the values, if they want,

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java [hunk 2]
```diff
@@ -74,6 +74,7 @@
 import org.nuxeo.ecm.core.schema.DocumentType;
 import org.nuxeo.ecm.core.schema.SchemaManager;
 import org.nuxeo.ecm.core.schema.types.ComplexType;
+import org.nuxeo.ecm.core.schema.types.CompositeType;
```
Comment (troger) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java]: Why `CompositeType`? Any issue with using `DocumentType` as it's already done?

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java [hunk 3]
```diff
@@ -295,36 +298,58 @@ protected void doImport(CSVParser parser) {
...
+        String newParentPath = targetPath.removeLastSegments(1).toString();
+        boolean exists = options.getCSVImporterDocumentFactory().exists(session, newParentPath, name);
+
+        DocumentRef docRef = null;
+        String type = null;
+        if (exists) {
+            docRef = new PathRef(targetPath.toString());
+            type = session.getDocument(docRef).getType();
+        } else {
+            if (Boolean.FALSE.equals(isTypeColumnMissing(header))) {
```
Comment (troger) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java]: `if (!isTypeColumnMissing(header)`?

However, I would prefer to read something like:

    if (hasTypeColumn(header)) {
        type = record.get(CSV_TYPE_COL);
    }
    ....

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java [hunk 4]
```diff
@@ -295,36 +298,58 @@ protected void doImport(CSVParser parser) {
...
+        } else {
+            return createDocument(lineNumber, newParentPath, name, type, properties);
+        }
     }
 
     /**
      * @since 6.0
      */
-    protected Map<String, Serializable> computePropertiesMap(CSVRecord record, DocumentType docType,
+    protected Map<String, Serializable> computePropertiesMap(CSVRecord record, CompositeType compositeType,
```
Comment (troger) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java]: Same here, we should keep `DocumentType`.

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java [hunk 5]
```diff
@@ -438,17 +463,12 @@ protected DateFormat getDateFormat() {
...
-        name = targetPath.lastSegment();
-        String newParentPath = targetPath.removeLastSegments(1).toString();
-        DocumentRef docRef = new PathRef(targetPath.toString());
-        if (options.getCSVImporterDocumentFactory().exists(session, newParentPath, name, type, properties)) {
-            return updateDocument(lineNumber, docRef, properties);
-        } else {
-            return createDocument(lineNumber, newParentPath, name, type, properties);
+    protected boolean isTypeColumnMissing(Map<String, Integer> header) {
+        // transient field so may become null
+        if (isTypeColumnMissing == null) {
```
Comment (troger) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java]: Why not just setting `hasTypeColum` field in the `doImport` method when we retrieve the `headers` and avoid this method?

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterDocumentFactory.java (hunk 1 above)
Comment (ron1) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterDocumentFactory.java]: Currently, the CoreSession does not provide a reliable means by which these values could be used to accomplish the operation. However, I presume the importer itself could use these values locally to accomplish some de-duping of the CSV rows, for example. I will modify my patch per your suggestion.

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java (hunk 2 above)
Comment (ron1) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java]: The more general CompositeType makes the computePropertiesMap() method more extensible. For example, it allows the method to be passed a dynamic facet in addition to a document type. This makes it much easier for custom CSVImporterWork extensions to support dynamic facets, for example.

File: src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java (hunk 5 above)
Comment (ron1) [inline src/main/java/org/nuxeo/ecm/csv/CSVImporterWork.java]: I'll give that a shot.

**Pull request closed without merging**